### PR TITLE
Add timeout for running actions

### LINKED ISSUE #463: Add a way to cancel actions
From `act/registry.go:253`:
> // TODO: Add a way to cancel the action.

This can be achieved using `context.WithTimeout` or something similar, just like policy evaluation timeout.

### PR DESCRIPTION
**Pull request by Hamsajj**
# Ticket(s)

## Description
(This PR is in draft because it depends on https://github.com/gatewayd-io/gatewayd-plugin-sdk/pull/26)
This PR will close #463 by adding timeout for running action.

Timeout will be defined by two different configs. One on the act system level, which will act as a default timeout and is added in this PR. the other config is on Action level, which allows each action to have a different timeout and will override the default one.
That config is added in https://github.com/gatewayd-io/gatewayd-plugin-sdk/pull/26

If both of these values are zero, actions will run without any timeout.
## Related PRs
https://github.com/gatewayd-io/gatewayd-plugin-sdk/pull/26
## Development Checklist

- [x] I have added a descriptive title to this PR.
- [x] I have squashed related commits together.
- [x] I have rebased my branch on top of the latest main branch.
- [x] I have performed a self-review of my own code.
- [x] I have commented on my code, particularly in hard-to-understand areas.
- [x] I have added docstring(s) to my code.
- [x] I have made corresponding changes to the documentation (docs).
- [x] I have updated docs using `make gen-docs` command.
- [x] I have added tests for my changes.
- [x] I have signed all the commits.

## Legal Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/gatewayd-io/gatewayd/blob/main/CONTRIBUTING.md) document.
- [x] I have read and understood the [Code of Conduct](https://github.com/gatewayd-io/gatewayd/blob/main/CODE_OF_CONDUCT.md).
- [x] I have read and agreed to the [Apache CLA](https://www.apache.org/licenses/contributor-agreements.html) (required).
- [x] I have read and agreed to the [LICENSE](https://github.com/gatewayd-io/gatewayd/blob/main/LICENSE) (required).

### CODE REVIEW

**Comment by mostafa** [comment]
@Hamsajj I merged the changes on the SDK and you can rebase from `main` now. :pray:

File: act/registry.go [hunk 1]
```diff
@@ -229,41 +233,68 @@ func (r *Registry) Run(
...
+		timeout = time.Duration(action.Timeout) * time.Second
+	}
+	var ctx context.Context
+	var cancel context.CancelFunc
+	// if timeout is zero, then the context should not have timeout
+	if timeout > 0 {
+		ctx, cancel = context.WithTimeout(context.Background(), timeout)
+	} else {
+		ctx, cancel = context.WithCancel(context.Background())
+
```
Comment (mostafa) [inline act/registry.go:247]: Empty line:
```suggestion

```

File: act/registry_test.go [hunk 2]
```diff
@@ -456,11 +457,136 @@ func Test_Run_ActionNotExist(t *testing.T) {
...
+				time.Sleep(3 * time.Millisecond)
+			}
+			if test.expectedLog != "" {
+				assert.Contains(t, out.String(), test.expectedLog)
+			}
+		})
+	}
+}
+
+func createWaitActEntities(async bool) (
```
Comment (mostafa) [inline act/registry_test.go:550]: Nit: move this function to a separate file: `act/act_helpers_test.go`.

File: go.mod [hunk 3]
```diff
@@ -89,3 +89,5 @@ require (
 	google.golang.org/appengine v1.6.8 // indirect
 	google.golang.org/genproto/googleapis/rpc v0.0.0-20240314204616-9694c7771956 // indirect
 )
+
+replace github.com/gatewayd-io/gatewayd-plugin-sdk => ../gatewayd-plugin-sdk
```
Comment (mostafa) [inline go.mod:93]:
```suggestion

replace github.com/gatewayd-io/gatewayd-plugin-sdk => ../gatewayd-plugin-sdk
```

**Review by mostafa — COMMENTED** [review]
Hey @Hamsajj,

Thanks for your awesome contribution! :pray:

Overall LGTM! I added a few comments. Note that I updated the SDK, so you can rebase from the `main` branch and tidy the deps.

**Review by mostafa — APPROVED** [review]
:rocket:

File: act/registry.go (hunk 1 above)
Comment (Hamsajj) [reply act/registry.go]: good catch. forgot to ran `golangci-lint`. It is fixed now.

File: act/registry_test.go (hunk 2 above)
Comment (Hamsajj) [reply act/registry_test.go]: moved it to the file as you said

**Pull request merged**